pysoundcomparisons/__main__.py

```diff
@@ -202,88 +200,96 @@ def downloadSoundFiles(args, out_path=os.path.join(os.getcwd(), "sound"), db_nee
                 elif k == 2: # FilePathPart + WordID
-                    if i in all_catalog_keys:
+                    if i in catalog:
                         desired_keys.add(i)
                     else:
-                        print("%s unknown - will be ignored" % (i), flush=True)
+                        args.log.warning("%s unknown - will be ignored" % (i))
                 else:
-                    print("%s unknown - will be ignored" % (i), flush=True)
+                    args.log.warning("%s unknown - will be ignored" % (i))
     else:
         for i in args.args:
             k = len(re.split(r"_\d{3,}_", i))
             if k == 1: # FilePathPart only ?
-                found = False
-                i_ = i + "_"
-                for s in all_catalog_keys:
-                    if s.startswith(i_):
-                        desired_keys.add(s)
-                        found = True
-                if not found:
-                    print("%s unknown - will be ignored" % (i), flush=True)
+                new_keys = catalog.file_path_keys_with_prefix(i)
+                if len(new_keys) > 0:
+                    desired_keys.update(new_keys)
+                else:
+                    args.log.warning("%s unknown - will be ignored" % (i))
             elif k == 2: # FilePathPart + WordID
-                if i in all_catalog_keys:
+                if i in catalog:
                     desired_keys.add(i)
                 else:
-                    print("%s unknown - will be ignored" % (i), flush=True)
+                    args.log.warning("%s unknown - will be ignored" % (i))
             else:
-                print("%s unknown - will be ignored" % (i), flush=True)
+                args.log.warning("%s unknown - will be ignored" % (i))
 
     # download all desired sound files from CDSTAR
+    if "CDSTAR_URL" in os.environ:
+        service_url = os.environ["CDSTAR_URL"]
+    else:
+        service_url = "http://cdstar.shh.mpg.de"
     cur_sffolder = ""
     if not os.path.exists(out_path):
         os.makedirs(out_path)
     for pth in sorted(desired_keys):
+        # check for available bitstreams for desired mimitypes
+        bs_objs = catalog.matching_bitstreams_for_mimetypes(
+            pth, [catalog.valid_mimetypes[ext] for ext in desired_ext])
+        if len(bs_objs) == 0:
+            #falling back to first bitstream if any
+            try:
+                bs_objs = [catalog[pth].bitstreams[0]]
+            except:
+                args.log.warning("No sound file for %s found" % (pth))
+                continue
         # get folder name out of pth
         p = re.split(r"_\d{3,}_", pth)
         if len(p) != 2:
-            print("Invalid key %s - will be ignored" % (s))
-            next
+            args.log.warning("Invalid key %s - will be ignored" % (s))
+            continue
         sffolder = p[0]
         if sffolder != cur_sffolder:
```

Comment:
yes - but what happens if a desired_key cannot split into two strings - the current approach is ignoring that case

Reply:
But at this point in the processing, only valid keys should be in `desired_keys`, right?

Reply:
yes if we assume that this key structure (file paths names) does not change in future

Reply:
I'd also turn the splitting into a function (presumably a method of `MediaCatalog`), so this function could return `None` for non-splittable keys, and items where `folder is None` could be ignored.

Reply:
AFAICT the code assumes this key structure in quite a few places, so I'd say that we can take this as a given; or the other way round: when we change the data structure, we'll have to change the code.

From my experience future proofing code by making it always do something, rather than breaking when assumptions are not met, does not really help.

Reply:
ok - but then I'd go with a new class `SoundfilePath` in MediaCatalog with attributes: `extension` (if given), `mimetype` (if given and comes from MediaCatalog's mime types), `member_of`(the language as FilePathPart = folder name) and maybe a method `is_valid` (calculate against file `valid_soundfilepaths.txt` or database)

Reply:
maybe as subclass of path lib.Path

class SoundfilePath(Path):

Reply:
Hm. That might be overkill. In particular considering that we want to use this as `dict` keys, so it needs to be hashable as well. Unfortunately, also, sub-classing `pathlib.Path` isn't straightforward. Maybe one could subclass `str`.

Reply:
Ok, maybe this would do:
```python
class SoundfileName(str):
    @property
    def path(self):
        return pathlib.Path(self)
```
Extension could then be accessed as `sf.path.suffix`.

Reply:
subclassing `str` would be nice since we would need this `member_of`very often and if something will change we only have to adapt it once

Reply:
We could even throw in
```python
    def __new__(cls, content):
        if not re.search('_\d{3,}_', content):
            raise ValueError()
      return str.__new__(cls, content)
```
for validation.

Reply:
👍

Reply:
I would call the property `variety`, though. `member_of` sounds pretty opaque to me.